msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java

```diff
@@ -162,4 +171,39 @@ void acquireToken(@NonNull final Activity activity,
         void onError(final MsalException exception);
     }
 
+    /**
+     * Callback object used in Device Code Flow.
+     * This callback provides the following methods for communicating with the protocol.
+     *                 1). Receiving authentication information (user_code, verification_uri, and instruction message)
+     *                 via {@link DeviceCodeFlowCallback#onUserCodeReceived(String, String, String)}.
+     *                 2). Receiving a successful authnetication result containing a fresh access token
+     *                 via {@link DeviceCodeFlowCallback#onTokenReceived(AuthenticationResult)}.
+     *                 3). Receiving an exception detailing what went wrong in the protocol
+     *                 via {@link DeviceCodeFlowCallback#onError(MsalServiceException)}.
+     */
+    interface DeviceCodeFlowCallback {
+        /**
+         * Invoked to display verification uri, user code, and instruction message during device code flow.
+         *
+         * @param vUri verification uri
+         * @param userCode user code
+         * @param message instruction message
+         */
+        void onUserCodeReceived(@NonNull String vUri, @NonNull String userCode, @NonNull String message);
+
+        /**
+         * Invoked once succeed and pass the result object.
+         *
+         * @param authResult the authentication result
+         */
+        void onTokenReceived(AuthenticationResult authResult);
+
+        /**
+         * Invoked once exception thrown.
+         *
+         * @param error error exception
+         */
+        void onError(MsalServiceException error);
```

Comment:
Can we get other types of Exceptions here? How about any `MsalException`?

Reply:
It originally was `MsalException`, but I think the command class will only be throwing `MsalServiceException`.

Reply:
I think I will change it back to MsalException just to be safe.